influxdb3_write/src/parquet_cache/mod.rs

```diff
@@ -108,59 +115,208 @@ struct CacheValue {
+        Self {
+            capacity,
+            used: AtomicUsize::new(0),
+            max_free_amount: (capacity as f64 * max_free_percentage).floor() as usize,
+            map: RwLock::new(IndexMap::new()),
+        }
     }
 
-    fn is_success(&self) -> bool {
-        matches!(self, CacheEntry::Success(_))
+    /// Get an entry in the cache or `None` if there is not an entry
+    ///
+    /// If the entry is in `Fetching` state, then this will await the etry having been fetched.
+    async fn get(&self, path: &Path) -> Option<Arc<CacheValue>> {
+        let map = self.map.read().await;
+        let entry = map.get(path)?;
+        match &entry.state {
+            CacheEntryState::Fetching(tx) => {
+                let mut rx = tx.subscribe();
+                // TODO(trevor): is it possible that the sender has sent the result at this point?
+                // if so we need to check the rx first and maybe use more sophisticated state than
+                // an Option in the watch channel type.
+                rx.changed().await.ok()?;
+                let v = rx.borrow_and_update().clone();
+                v
+            }
+            CacheEntryState::Success(v) => {
+                entry.increment_hits();
+                Some(Arc::clone(v))
+            }
+        }
     }
 
-    fn keep(&self) -> bool {
-        self.is_fetching() || self.is_success()
+    /// Check if an entry in the cache is in process of being fetched or if it was already fetched
+    /// successfully
+    async fn path_already_fetched(&self, path: &Path) -> bool {
+        self.map.read().await.get(path).is_some()
     }
-}
 
-/// Implements the [`WeightScale`] trait to determine a [`CacheEntry`]'s size on insertion to
-/// the cache
-#[derive(Debug)]
-struct CacheEntryScale;
+    /// Insert a `Fetching` entry to the cache with a watcher
+    async fn set_fetching(&self, path: &Path) {
+        let (tx, _) = watch::channel(None);
+        let entry = CacheEntry {
+            state: CacheEntryState::Fetching(tx),
+            hits: AtomicI8::new(-1),
+        };
+        let additional = entry.size();
+        self.map.write().await.insert(path.clone(), entry);
+        self.used.fetch_add(additional, Ordering::SeqCst);
+    }
+
+    /// Update a `Fetching` entry to a `Success` entry in the cache
+    async fn set_success(&self, path: &Path, value: CacheValue) -> Result<(), anyhow::Error> {
+        match self.map.write().await.entry(path.clone()) {
+            Entry::Occupied(mut o) => {
+                let entry = o.get_mut();
+                let tx = match &entry.state {
+                    CacheEntryState::Fetching(tx) => tx.clone(),
+                    _ => bail!("attempted to set success state on a non-fetching cache entry"),
+                };
+                let cache_value = Arc::new(value);
+                entry.state = CacheEntryState::Success(Arc::clone(&cache_value));
+                entry.hits.store(1, Ordering::SeqCst);
```

Comment:
You could use the [receiver count](https://docs.rs/tokio/latest/tokio/sync/watch/struct.Sender.html#method.receiver_count) on the `watch::Sender` to account for the extra hits due to waiting requests here, vs. just storing a `1`